src/renderer/wgpu_renderer.rs

```diff
@@ -0,0 +1,292 @@
+        let instance = Instance::new(InstanceDescriptor::default());
+
+        let raw_display_handle = window.raw_display_handle();
+        let raw_window_handle = window.raw_window_handle();
+
+        let target = SurfaceTargetUnsafe::RawHandle {
+            raw_display_handle: match raw_display_handle {
+                raw_window_handle::RawDisplayHandle::AppKit(_) => {
+                    raw_window_handle_06::RawDisplayHandle::AppKit(AppKitDisplayHandle::new())
+                }
+                raw_window_handle::RawDisplayHandle::Xlib(handle) => {
+                    raw_window_handle_06::RawDisplayHandle::Xlib(XlibDisplayHandle::new(
+                        NonNull::new(handle.display),
+                        handle.screen,
+                    ))
+                }
+                raw_window_handle::RawDisplayHandle::Xcb(handle) => {
+                    raw_window_handle_06::RawDisplayHandle::Xcb(XcbDisplayHandle::new(
+                        NonNull::new(handle.connection),
+                        handle.screen,
+                    ))
+                }
+                raw_window_handle::RawDisplayHandle::Windows(_) => {
+                    raw_window_handle_06::RawDisplayHandle::Windows(WindowsDisplayHandle::new())
+                }
+                _ => todo!(),
+            },
+            raw_window_handle: match raw_window_handle {
+                raw_window_handle::RawWindowHandle::AppKit(handle) => {
+                    raw_window_handle_06::RawWindowHandle::AppKit(AppKitWindowHandle::new(
+                        NonNull::new(handle.ns_view).unwrap(),
+                    ))
+                }
+                raw_window_handle::RawWindowHandle::Xlib(handle) => {
+                    raw_window_handle_06::RawWindowHandle::Xlib(XlibWindowHandle::new(
+                        handle.window,
+                    ))
+                }
+                raw_window_handle::RawWindowHandle::Xcb(handle) => {
+                    raw_window_handle_06::RawWindowHandle::Xcb(XcbWindowHandle::new(
+                        NonZeroU32::new(handle.window).unwrap(),
+                    ))
+                }
+                raw_window_handle::RawWindowHandle::Win32(handle) => {
+                    // will this work? i have no idea!
+                    let mut raw_handle =
+                        Win32WindowHandle::new(NonZeroIsize::new(handle.hwnd as isize).unwrap());
+
+                    raw_handle.hinstance = handle
+                        .hinstance
+                        .is_null()
+                        .then(|| NonZeroIsize::new(handle.hinstance as isize).unwrap());
+
+                    raw_window_handle_06::RawWindowHandle::Win32(raw_handle)
+                }
+                _ => todo!(),
+            },
+        };
+
+        let surface = unsafe { instance.create_surface_unsafe(target) }.unwrap();
+        let configuration = WgpuConfiguration::default();
```

Comment:
This should be a user-configurable parameter inside the `WindowOpenOptions` struct.

Reply:
there's not really an easy way to do that, because the device_descriptor field doesn't implement send. the compiler gets really mad at me if i try do give it to the closure. maybe wrapping it in another arc will fix it? but probably not

Reply:
an option i thought of is passing a closure that returns the options to get around the trait bounds, but doing that as a struct field means generics, which i don't want to deal with trying to feature gate that

Reply:
Hmm, I see. Does it work if we redefine our own `WgpuConfiguration` struct like this?:
```rust
pub struct WgpuConfiguration {
    pub supported_backends: Backends,
    pub device_descriptor: Arc<dyn Fn(&Adapter) -> DeviceDescriptor<'static> + Send>,
    pub present_mode: PresentMode,
    pub desired_maximum_frame_latency: Option<u32>,
    pub power_preference: PowerPreference,
    pub on_surface_error: Arc<dyn Fn(SurfaceError) -> SurfaceErrorAction + Send>,
}
```

Reply:
i can try, i'll need to think about how to convert between the 2, i'll experiment with it

Reply:
i found a way to do it, the only gross thing is i had to copy paste egui's default implementation which bodes badness if they ever get out of sync